File: Inverted-Index/src/tests/SearchEngineTest.java
package tests;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Map;
import java.util.HashMap;

import static org.mockito.Mockito.*;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;
import org.junit.Assert;
import org.junit.Before;

import search.InvertedIndex;
import search.QueryParser;
import search.SearchEngine;
import file_handler.Doc;
import keywords.Exclude;
import keywords.Ordinary;
import keywords.Union;

@RunWith(MockitoJUnitRunner.class)
public class SearchEngineTest {

    private final Doc DOC1;
    private final Doc DOC2;
    private final Doc DOC3;

    @Mock
    QueryParser parser;
    @Mock
    InvertedIndex index;
    @Mock
    Ordinary ordinary;
    @Mock
    Union union;
    @Mock
    Exclude exclude;

    public SearchEngineTest() {
        DOC1 = new Doc("57110", new ArrayList<>());
        DOC2 = new Doc("59483", new ArrayList<>());
        DOC3 = new Doc("59519", new ArrayList<>());
    }

    private Set<Doc> buildHashSet(boolean doc1Exists, boolean doc2Exists, boolean doc3Exists) {
        var set = new HashSet<Doc>();
        if(doc1Exists) set.add(DOC1);
        if(doc2Exists) set.add(DOC2);
        if(doc3Exists) set.add(DOC3);

        return set;
    }
    
    private Set<Doc> mockDocs(Map<Keyword,boolean[]> keywordDocs) {

        var keywordsMock = keywordDocs.keySet();
        when(parser.parseQuery(anyString())).thenReturn(keywordsMock);

        for(var pair : keywordDocs) {
            var includeArray = pair.getValue();
            when(pair.getKey().operate(any(), any())).thenReturn(
                buildHashSet(includeArray[0],includeArray[1],includeArray[2])
            );
        }

        var docs = engine.search(parser.parseQuery(""));
    }

    @Before
    public void mockResources() {

        var engine = new SearchEngine(index);

        doNothing().when(index.get(any()));
        
        doNothing().when(ordinary.getWord());
        doNothing().when(union.getWord());
        doNothing().when(exclude.getWord());
    }

    @Test
    public void unionWithExcludeTest() {
        var keywordDocs = new HashMap<Keyword,boolean[]>() {
            {
                put(ordinary, {false,true,false});
                put(union, {true,true,false});
                put(exclude, {true,false,true});
            }
        };
        
        var docs = mockDocs(keywordDocs);
        var expectedDocs = buildHashSet(false,true,false);
        Assert.assertEquals(expectedDocs, docs);
    }

    @Test
    public void includeAllTest() {
        var keywordDocs = new HashMap<Keyword,boolean[]>() {
            {
                put(ordinary, {false,false,false});
                put(union, {true,true,true});
            }
        };
        
        var docs = mockDocs(keywordDocs);
        var expectedDocs = buildHashSet(true,true,true);
        Assert.assertEquals(expectedDocs, docs);
    }

    @Test
    public void excludeAllTest() {
        
        var keywordDocs = new HashMap<Keyword,boolean[]>() {
            {
                put(ordinary, {true,false,true});
                put(exclude, {true,true,true});
            }
        };
        
        var docs = mockDocs(keywordDocs);
        var expectedDocs = buildHashSet(false,false,false);
        Assert.assertEquals(expectedDocs, docs);
    }

    @Test
    public void simpleSearchTets() {
        var keywordDocs = new HashMap<Keyword,boolean[]>() {
            {
                put(ordinary, {true,true,false});
            }
        };
        
        var docs = mockDocs(keywordDocs);
        var expectedDocs = buildHashSet(true,true,false);
        Assert.assertEquals(expectedDocs, docs);
    }
}

File: Phase5-TDD/InvertedIndex.Test/SearchEngineTest.cs
using System.Linq;
using System.Collections.Generic;
using Xunit;
using Moq;
using InvertedIndex.Search;
using InvertedIndex.FileHandler;
using InvertedIndex.Keywords;

namespace InvertedIndex.Test
{
    public class SearchEngineTest 
    {
        private readonly Doc DOC1;
        private readonly Doc DOC2;
        private readonly Doc DOC3;

        Mock<QueryParser> _parserMock;
        Mock<Search.InvertedIndex> _indexMock;
        Mock<Keyword> _ordinaryMock;
        Mock<Keyword> _unionMock;
        Mock<Keyword> _excludeMock;

        SearchEngine _engine;

        public SearchEngineTest()
        {
            DOC1 = new Doc("57110", new List<string>());
            DOC2 = new Doc("59483", new List<string>());
            DOC3 = new Doc("59519", new List<string>());

            _indexMock = new Mock<Search.InvertedIndex>();
            _indexMock.Setup(i => i.Get(It.IsAny<string>())).Verifiable();
            _engine = new SearchEngine(_indexMock.Object);

            _ordinaryMock = new Mock<Keyword>();
            _excludeMock = new Mock<Keyword>();
            _unionMock = new Mock<Keyword>();
            _ordinaryMock.Setup( o => o.GetWord()).Verifiable();
            _excludeMock.Setup( o => o.GetWord()).Verifiable();
            _unionMock.Setup( o => o.GetWord()).Verifiable();
        }

        private HashSet<Doc> buildHashSet(bool doc1Exists, bool doc2Exists, bool doc3Exists)
        {
            var set = new HashSet<Doc>();
            if(doc1Exists) set.Add(DOC1);
            if(doc2Exists) set.Add(DOC2);
            if(doc3Exists) set.Add(DOC3);

            return set;
        }

        private HashSet<Doc> mockDocs(IDictionary<Mock<Keyword>,bool[]> keywordDocs)
        {

            var keywordsMock = keywordDocs.Keys.Select(k => k.Object).ToList();
            _parserMock = new Mock<QueryParser>();
            _parserMock.Setup(p => p.ParseQuery(It.IsAny<string>())).Returns(keywordsMock);

            foreach(var pair in keywordDocs)
            {
                var includeArray = pair.Value;
                pair.Key.Setup(k => k.Operate(It.IsAny<HashSet<Doc>>(), It.IsAny<HashSet<Doc>>()))
                    .Returns(buildHashSet(includeArray[0], includeArray[1], includeArray[2]));
            }

            return _engine.Search(_parserMock.Object.ParseQuery(""));
        }

        [Fact]
        public void unionWithExcludeTest()
        {
            var keywordDocs = new Dictionary<Mock<Keyword>,bool[]>();
            keywordDocs.Add(_ordinaryMock, new bool[] {false,true,false});
            keywordDocs.Add(_unionMock, new bool[] {true,true,false});
            keywordDocs.Add(_excludeMock, new bool[] {false,true,false});
            
            var docs = mockDocs(keywordDocs);
            var expectedDocs = buildHashSet(false,true,false);
            Assert.Equal(expectedDocs, docs);
        }

        [Fact]
        public void includeAllTest()
        {
            var keywordDocs = new Dictionary<Mock<Keyword>,bool[]>();

            keywordDocs.Add(_ordinaryMock, new bool[] {false,false,false});
            keywordDocs.Add(_unionMock, new bool[] {true,true,true});
            
            
            var docs = mockDocs(keywordDocs);
            var expectedDocs = buildHashSet(true,true,true);
            Assert.Equal(expectedDocs, docs);
        }

        [Fact]
        public void excludeAllTest()
        {    
            var keywordDocs = new Dictionary<Mock<Keyword>,bool[]>();
            keywordDocs.Add(_ordinaryMock, new bool[] {true,false,true});
            keywordDocs.Add(_excludeMock, new bool[] {false,false,false});
            
            var docs = mockDocs(keywordDocs);
            var expectedDocs = buildHashSet(false,false,false);
            Assert.Equal(expectedDocs, docs);
        }

        [Fact]
        public void simpleSearchTets()
        {
            var keywordDocs = new Dictionary<Mock<Keyword>,bool[]>();

            keywordDocs.Add(_ordinaryMock, new bool[] {true,true,false});

            
            var docs = mockDocs(keywordDocs);
            var expectedDocs = buildHashSet(true,true,false);
            Assert.Equal(expectedDocs, docs);
        }
    }
}